payjoin/src/uri.rs

```diff
@@ -274,10 +327,91 @@ enum InternalPjParseError {
     UnsecureEndpoint,
 }
 
+/// Builder for `bip21::Uri` with `PayjoinParams`.
+///
+/// `amount` parameter is optional and can be changed with `[PjUriBuilder::amount]`.
+///
+/// `ohttp_config` parameter is required only for v2 payjoin.
+pub struct PjUriBuilder {
+    /// Address you want to receive funds to.
+    ///
+    /// Must be a valid bitcoin address.
+    address: Address,
+    /// Payjoing endpoint url listening for payjoin requests.
+    ///
+    /// Must be a valid url that can be parsed
+    /// with `[Payjoin::Url::parse]`.
+    pj_endpoint: Url,
+    /// Amount you want to receive.
+    ///
+    /// If `None` the amount will be left unspecified.
+    amount: Option<Amount>,
+    #[cfg(feature = "v2")]
+    /// Config for ohttp.
+    ///
+    /// `[PjUriBuilder::decode_ohttp_config]` can be used to convert
+    /// base64 string to `ohttp::KeyConfig`.
+    ///
+    /// Required only for v2 payjoin.
+    ohttp_config: ohttp::KeyConfig,
+}
+
+impl PjUriBuilder {
+    /// Create a new `PjUriBuilder` with required parameters.
+    pub fn new(
+        address: Address,
+        pj_endpoint: Url,
+        #[cfg(feature = "v2")] ohttp_config: ohttp::KeyConfig,
+    ) -> Self {
+        Self {
+            address,
+            pj_endpoint,
+            amount: None,
+            #[cfg(feature = "v2")]
+            ohttp_config,
+        }
+    }
+
+    /// Set the amount you want to receive.
+    pub fn amount(mut self, amount: Option<bitcoin::Amount>) -> Self {
+        self.amount = amount;
+        self
+    }
+
+    /// Build a `bip21::Uri` with `PayjoinParams`
+    pub fn build<'a>(self) -> Result<bip21::Uri<'a, NetworkChecked, Payjoin>, PjParseError> {
```

Comment:
I believe we can do the fallable checks in the setters or better yet make them infallable so that this function does not need to return a `Result`